### ISSUE TITLE
Ubuntu 14.04 LTS libOpenCvSharpExtern Issue

### ISSUE THREAD

**Issue by jbob182** [shimat/opencvsharp#85]
I'm having difficulties getting OpenCvSharp.CPlusPlus to load the library libOpenCvSharpExtern.so 

My environment: Ubuntu 14.04 LTS, x64, Mono JIT compiler version 3.2.8

When executing the mono program under normal log levels I get:
```
System.DllNotFoundException: libOpenCvSharpExtern.so
  at (wrapper managed-to-native) OpenCvSharp.CPlusPlus.NativeMethods:core_Mat_sizeof ()
  at OpenCvSharp.CPlusPlus.NativeMethods.TryPInvoke () [0x00000] in <filename unknown>:0  ---> System.DllNotFoundException: libOpenCvSharpExtern.so
  at (wrapper managed-to-native) OpenCvSharp.CPlusPlus.NativeMethods:core_Mat_sizeof ()
  at OpenCvSharp.CPlusPlus.NativeMethods.TryPInvoke () [0x00000] in <filename unknown>:0 
  --- End of inner exception stack trace ---
  at OpenCvSharp.CPlusPlus.NativeMethods.TryPInvoke () [0x00000] in <filename unknown>:0 
  at OpenCvSharp.CPlusPlus.NativeMethods..cctor () [0x00000] in <filename unknown>:0 
  --- End of inner exception stack trace ---
  at OpenCvSharp.CPlusPlus.Window..ctor (System.String name, WindowMode flags, OpenCvSharp.CPlusPlus.Mat image) [0x00000] in <filename unknown>:0 
  at OpenCvSharp.CPlusPlus.Window..ctor (System.String name) [0x00000] in <filename unknown>:0 
  at Test.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0 
```
However, if I run my app with MONO_LOG_LEVEL=debug then I get the following:
```
Mono: DllImport error loading library '/home/jbob/workspace/Test/Test/bin/Debug/libOpenCvSharpExtern.so': '/home/jbob/workspace/Test/Test/bin/Debug/libOpenCvSharpExtern.so: undefined symbol: _ZTIN2cv4SIFTE'.
```

Any advice you could provide would be greatly appreciated.

**Comment by AlvarezAti90** [comment]
I have the very same problem under Debian with Mono JIT compiler version 2.10.8.1 (Debian 2.10.8.1-8) .

Have you find any solution for this issue?

Thanks for any advice.

**Comment by Bohdan-Sydor** [comment]
Same issue, please help.

**Comment by ewenger** [comment]
I had the same problem and wouldn't have gotten anywhere without the MONO_LOG_LEVEL advice above.  I was able to get mine to work on Ubuntu 14.04 by adding -lopencv_<libname> commands to the linker parameters for libOpenCvSharpExtern.so.  This creates secondary dependencies for that shared object on the OpenCV libs and will pull them in as expected.  Try running ldd on libOpenCvSharpExtern.so and you'll see no OpenCV dependencies.  I found this page helpful: http://www.kaizou.org/2015/01/linux-libraries/

**Comment by Bohdan-Sydor** [comment]
**ewenger**, thank you so much, it works! For some reason in cmake file `target_link_libraries` is specified only for WIN32. If needed, I can commit fix which helped me (but I'm very new to C++ and cmake)

**Comment by ewenger** [comment]
I'm glad to hear it and I appreciate the real fix.  I didn't investigate the CMake files after I got my first sample program running.  I'll try out your recommendation today.

**Comment by yukoba** [comment]
Thank you everyone for this information.
I created a pull request https://github.com/shimat/opencvsharp/pull/109 based on this information.

**Comment by Nalgene01** [comment]
Hello,

I understand this has been resolved, I too am using Ubuntu 14.04 with Monodevelop 5.10. I am a Windows Guy **very new** to linux and Ubuntu.  I am getting the same error with this version of OpenCvSharp2.  2.4.10.20160316.  When I look at my dependencies of my .so I see no reference to OpecCV.  I installed openCV 2.4.10 as instructed.  On April 17 ewenger posted the solution and a link to documentation. Here is his response: I was able to get mine to work on Ubuntu 14.04 by adding -lopencv_ commands to the linker parameters for libOpenCvSharpExtern.so. 

Question: Where and how do I add the linker setting.  Can someone provide what modifications are needed to do this in the cmakefile or a link to documentation on what this means?

Thanks for any help,
Nalgene

System.DllNotFoundException: libOpenCvSharpExtern.so
  at (wrapper managed-to-native) OpenCvSharp.CPlusPlus.NativeMethods:core_Mat_sizeof ()
  at OpenCvSharp.CPlusPlus.NativeMethods.TryPInvoke () [0x00014] in <filename unknown>:0 

Unhandled Exception:
System.TypeInitializationException: The type initializer for 'OpenCvSharp.CPlusPlus.Mat' threw an exception. ---> System.TypeInitializationException: The type initializer for 'OpenCvSharp.CPlusPlus.NativeMethods' threw an exception. ---> OpenCvSharp.OpenCvSharpException: libOpenCvSharpExtern.so
*** An exception has occurred because of P/Invoke. ***
Please check the following:
(1) OpenCV's DLL files exist in the same directory as the executable file.
(2) Visual C++ Redistributable Package has been installed.
(3) The target platform(x86/x64) of OpenCV's DLL files and OpenCvSharp is the same as your project's.

System.DllNotFoundException: libOpenCvSharpExtern.so
  at (wrapper managed-to-native) OpenCvSharp.CPlusPlus.NativeMethods:core_Mat_sizeof ()
  at OpenCvSharp.CPlusPlus.NativeMethods.TryPInvoke () [0x00014] in <filename unknown>:0  ---> System.DllNotFoundException: libOpenCvSharpExtern.so
  at (wrapper managed-to-native) OpenCvSharp.CPlusPlus.NativeMethods:core_Mat_sizeof ()
  at OpenCvSharp.CPlusPlus.NativeMethods.TryPInvoke () [0x00014] in <filename unknown>:0 
  --- End of inner exception stack trace ---
  at OpenCvSharp.CPlusPlus.NativeMethods.TryPInvoke () [0x00038] in <filename unknown>:0 
  at OpenCvSharp.CPlusPlus.NativeMethods..cctor () [0x00006] in <filename unknown>:0

**Comment by fedorkov** [comment]
Looks like the last change of `src/OpenCvSharpExtern/CMakeLists.txt` ([replaced master from OpenCV2 to OpenCV3](https://github.com/shimat/opencvsharp/commit/07d26125061704868eaa130a9aed5861d8851797#diff-9eb5154f14db1b6190a84e234cdbf5d2)) broke it again. I tried to append
```
file(GLOB OPENCV_SO_FILES "/usr/lib/libopencv_*.so.3.1.0")
target_link_libraries(OpenCvSharpExtern PRIVATE ${OPENCV_SO_FILES})
```
to this file, and I think it worked. Not sure though, because when I try to run [this sample](https://github.com/shimat/opencvsharp#usage) I get `System.EntryPointNotFoundException: imgcodecs_imread`.
